src/librustdoc/html/highlight.rs

```diff
@@ -134,151 +135,321 @@ fn can_merge(class1: Option<Class>, class2: Option<Class>, text: &str) -> bool {
     }
 }
 
+#[derive(Debug)]
+struct ClassInfo {
+    class: Class,
+    /// Set to true only when an item was written inside this tag.
+    open: bool,
+    /// Set to true when leaving the current item. The closing tag will be
+    /// written if:
+    ///
+    /// 1. `self.open` is true
+    /// 2. Only when the first non-mergeable item is pushed.
+    pending_exit: bool,
+    /// If `true`, it means it's `</a>`, otherwise it's `</span>`.
+    link_closing_tag: bool,
+}
+
+impl ClassInfo {
+    fn new(class: Class, pending_exit: bool) -> Self {
+        Self { class, open: pending_exit, pending_exit, link_closing_tag: false }
+    }
+
+    fn close_tag<W: Write>(&self, out: &mut W) {
+        if self.open {
+            if self.link_closing_tag {
+                out.write_str("</a>").unwrap();
+            } else {
+                out.write_str("</span>").unwrap();
+            }
+        }
+    }
+}
+
+/// This represents the stack of HTML elements. For example a macro expansion
+/// will contain other elements which might themselves contain other elements
+/// (like macros).
+///
+/// This allows to easily handle HTML tags instead of having a more complicated
+/// state machine to keep track of which tags are open.
+#[derive(Debug)]
+struct ClassStack {
+    open_classes: Vec<ClassInfo>,
+}
+
+impl ClassStack {
+    fn new() -> Self {
+        Self { open_classes: Vec::new() }
+    }
+
+    fn enter_elem<W: Write>(
+        &mut self,
+        out: &mut W,
+        href_context: &Option<HrefContext<'_, '_>>,
+        new_class: Class,
+        pending_exit: bool,
+    ) {
```

Comment:
```suggestion
        new_class: Class,
        close_on_next_elem: bool,
    ) {
```

not sure if maybe this is more clear?

Reply:
Not really? It can be next item (if there is one) or it can be in hundreds of items if they are compatible/mergeable/the same.